### PR TITLE
Added python wrappers

### PR DESCRIPTION
**Pull request by OrkoHunter**
This adds python wrappers and tests.

### CODE REVIEW

**Comment by OrkoHunter** [comment]
I had to rename `types.py` to avoid conflicts with the python built-in module `types`.

File: metis.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+from decorators import convert_exceptions
+from _types import MetisError
+
+try:
+    from _metis import node_nd, part_graph, compute_vertex_separator
+    from _types import MetisOptions
+    __all__ = ['node_nested_dissection', 'partition', 'vertex_separator',
+               'MetisOptions']
+except ImportError:
+    __all__ = []
```
Comment (ysitu) [inline metis.py]: After a success installation, these imports should just work. Let's not hide any errors.

File: decorators.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            exc = e
+        gvars = globals()
+        lvars = locals()
+        lvars['convert_type'] = convert_type
+        if version_info[0] >= 3:
+            code = 'raise convert_type(exc).with_traceback(exc.__traceback__)'
+        else:
+            code = 'raise convert_type, (exc,), exc_info()[2]'
+        _exec(code, gvars, lvars)
+    return _convert_exceptions
```
Comment (ysitu) [inline decorators.py]: Add new line.

File: functions.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+try:
+    from _metis import *
+    __all__ = ['part_graph', 'node_nd', 'compute_vertex_separator']
+except ImportError:
+    pass
```
Comment (ysitu) [inline functions.py]: Is this file necessary?

File: metis.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+    """
+    if len(G) == 0:
+        return []
+
+    vwgt = [G.node[u].get(weight, 1) for u in G]
+    if all(w == 1 for w in vwgt):
+        vwgt = None
+
+    if G.is_directed() or G.is_multigraph():
+        G = nx.Graph(G.edges_iter())
```
Comment (ysitu) [inline metis.py]: `G.edges()`

File: functions.py (hunk 3 above)
Comment (OrkoHunter) [inline functions.py]: I think no. I guess it might be added to give output for the functions metis provide as `metis.functions.__all__`.
But that can be read from `metis.py` too. I'll remove it.

File: tests/test_metis.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    xadj = list(range(0, 2 * n + 1, 2))
+    adjncy = list(chain.from_iterable(zip(chain([n - 1], range(n - 1)),
+                                          chain(range(1, n), [0]))))
+    return xadj, adjncy
+
+
+class TestMetis:
+
+    @classmethod
+    def setupClass(cls):
```
Comment (ysitu) [inline tests/test_metis.py]: Remove this function.

File: tests/test_metis.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+
+def make_cycle(n):
+    xadj = list(range(0, 2 * n + 1, 2))
+    adjncy = list(chain.from_iterable(zip(chain([n - 1], range(n - 1)),
+                                          chain(range(1, n), [0]))))
+    return xadj, adjncy
+
+
+class TestMetis:
```
Comment (ysitu) [inline tests/test_metis.py]: `class TestMetis(object)`:

File: _types.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
+from numbers import Number
+
+from enums import *
+try:
+    from _metis import set_default_options
+except ImportError:
```
Comment (ysitu) [inline _types.py]: Remove `try... except`. `_metis` should just exist.

File: metis.py [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+from sys import exc_info
+
+import networkx as nx
+from networkx.utils import not_implemented_for
+
+from decorators import convert_exceptions
+from _types import MetisError, MetisOptions
+from _metis import node_nd, part_graph, compute_vertex_separator
+__all__ = ['node_nested_dissection', 'partition', 'vertex_separator',
+               'MetisOptions']
```
Comment (ysitu) [inline metis.py]: Indentation

File: metis.py (hunk 4 above)
Comment (ysitu) [inline metis.py]: `edges_iter()` is being removed. Please fix.

File: decorators.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+                raise
+            exc = e
+        gvars = globals()
+        lvars = locals()
+        lvars['convert_type'] = convert_type
+        if version_info[0] >= 3:
+            code = 'raise convert_type(exc).with_traceback(exc.__traceback__)'
+        else:
+            code = 'raise convert_type, (exc,), exc_info()[2]'
+        _exec(code, gvars, lvars)
```
Comment (ysitu) [inline decorators.py]: I have

```python
def _exec(code, gvars, lvars):
    exec(code, gvars, lvars)
```

in my repository. I forgot why I added this, but IIRC removing `_exec` broke something.

Comment (chebee7i) [inline decorators.py]: We'll definitely have some comments this time around? :)

Comment (OrkoHunter) [inline decorators.py]: A similar thing had happened with me, with `types.py`. Renaming it to `_types.py` worked for me but not the previous one because
```python
In [1]: import types
```
This always imported the built-in module `types` which I didn't know of.
```python
In [2]: types
Out[2]: <module 'types' from '/usr/lib/python2.7/types.pyc'>
```

Comment (OrkoHunter) [inline decorators.py]: @ysitu I added the `_exec` function. I might have missed it while taking from your repository.

Comment (chebee7i) [inline decorators.py]: Regarding `import types` failing, you should use relative imports if you want to avoid that---but a rename works just fine too.  `from . import types`.

Let's make sure we document why `_exec` is necessary.

Comment (OrkoHunter) [inline decorators.py]: > Let's make sure we document why _exec is necessary.

Yeah. We'll try it after we successfully build the library and find which `break` @ysitu is talking about.

> from . import types

This used to raise `Attempted relative import from non-package error`. Due to a namespace package, I guess we may not put `__init__.py` on the top.

**Comment by chebee7i** [comment]
>> from . import types

> This used to raise `Attempted relative import from non-package error`. Due to a namespace package, I guess we may not put `__init__.py` on the top.

That's a really unappealing "feature" of namespace packages. It seems like any sufficiently complicated namespace package would want the option to do relative imports, rather than having to hardcode the package name in absolute imports.  The use case of a namespace package consisting of more than one module isn't discussed in [PEP 420](https://www.python.org/dev/peps/pep-0420/)

And, would this even work in Python 3? Python 3 does not allow implicit relative imports at all. Also, note that the way to create namespace packages differs in 3.x than it does in 2.x.

Though I haven't dug into this, it seems like a namespace package is potentially more complicated then doing the simple approach:

```
# networkx/metis/__init__.py

from __future__ import absolute_import

# this is an absolute import to .../site-packages/python2.7/networkx_metis
from networkx_metis import *
```

Then users use it as:   `import networkx.metis as metis`.

**Comment by ysitu** [comment]
How about not using relative import at all? We know where this package will land after installation.

**Comment by OrkoHunter** [comment]
@ysitu That is good. Maybe we can change all the imports before the final release of `networkx-metis`. As of now, it's under development, we should just keep it so.

**Comment by chebee7i** [comment]
Yeah that's fine. Have you guys looked into how to do this in a 2/3 compatible way? Would we just use `pkgutil` or would we use setuptools?

**Comment by chebee7i** [comment]
And I do realize this is minutiae compared to the larger compilation issues you are focusing on now. So feel free to ignore my comments until they become relevant.

**Comment by OrkoHunter** [comment]
@chebee7i No, I'm reading things about it in particular and I think `setuptools` would be the way to go. However, we'll discuss about it.

File: _types.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
+from numbers import Number
+
+from enums import *
+from _metis import set_default_options
```
Comment (OrkoHunter) [inline _types.py]: This seems to create a problem of cyclic import. `_metis` imports `_types` and here `_types` import `_metis` back.

Comment (ysitu) [inline _types.py]: What about `import _metis` and use `_metis.set_default_options(self)` below?

Comment (ysitu) [inline _types.py]: BTW, should we consider consistently using `import some_module` and `some_module.some_function` instead of `from some_module import some_function`?

Comment (OrkoHunter) [inline _types.py]: This diff works.

```diff
diff --git a/_metis.pyx b/_metis.pyx
index 61cccd3..fc140c7 100644
--- a/_metis.pyx
+++ b/_metis.pyx
@@ -4,7 +4,7 @@ from os import dup, dup2, fdopen
 from sys import stdout
 from tempfile import TemporaryFile
 from _api cimport *
-from ._types import MetisError
+import _types
 
 __all__ = ['part_graph', 'node_nd', 'compute_vertex_separator']
 
@@ -134,7 +134,7 @@ cdef void convert_graph(xadj, adjncy, idx_t *nvtxs_ptr, idx_t **_xadj_ptr,
 
 cdef void check_result(int result, msg) except *:
     if result != METIS_OK:
-        raise MetisError(msg)
+        raise _types.MetisError(msg)
 
 
 def part_graph(xadj, adjncy, nparts, vwgt=None, vsize=None, adjwgt=None,
diff --git a/_types.py b/_types.py
index 0f4339a..7001826 100644
--- a/_types.py
+++ b/_types.py
@@ -1,7 +1,6 @@
 from numbers import Number
 
-from enums import *
-from _metis import set_default_options
+import _metis
 
 __all__ = ['MetisOptions', 'MetisError']
 
@@ -9,7 +8,7 @@ __all__ = ['MetisOptions', 'MetisError']
 class MetisOptions(object):
 
     def __init__(self):
-        set_default_options(self)
+        _metis.set_default_options(self)
 
     @property
     def ptype(self):
diff --git a/metis.py b/metis.py
index 0dffe9b..3fa87be 100644
--- a/metis.py
+++ b/metis.py
@@ -14,11 +14,12 @@ from sys import exc_info
 import networkx as nx
 from networkx.utils import not_implemented_for
 
-from decorators import convert_exceptions
-from _types import MetisError, MetisOptions
-from _metis import node_nd, part_graph, compute_vertex_separator
+from enums import *
+import decorators
+import _types
+import _metis
 __all__ = ['node_nested_dissection', 'partition', 'vertex_separator',
-           'MetisOptions']
+           '_types.MetisOptions']
 
 
 @contextmanager
@@ -46,7 +47,7 @@ def _convert_graph(G):
     return xadj, adjncy
 
 
-@convert_exceptions(nx.NetworkXError, (ValueError, TypeError, MetisError))
+@decorators.convert_exceptions(nx.NetworkXError, (ValueError, TypeError, _types.MetisError))
 def node_nested_dissection(G, weight='weight', options=None):
     """Compute a node ordering of a graph that reduces fill when the Laplacian
     matrix of the graph is LU factorized. The algorithm aims to minimize the
@@ -88,7 +89,7 @@ def node_nested_dissection(G, weight='weight', options=None):
     xadj, adjncy = _convert_graph(G)
 
     with _zero_numbering(options):
-        perm = node_nd(xadj, adjncy, vwgt, options)[0]
+        perm = _metis.node_nd(xadj, adjncy, vwgt, options)[0]
 
     nodes = list(G)
     perm = [nodes[i] for i in perm]
@@ -98,7 +99,7 @@ def node_nested_dissection(G, weight='weight', options=None):
 
 @not_implemented_for('directed')
 @not_implemented_for('multigraph')
-@convert_exceptions(nx.NetworkXError, (ValueError, TypeError, MetisError))
+@decorators.convert_exceptions(nx.NetworkXError, (ValueError, TypeError, _types.MetisError))
 def partition(G, nparts, node_weight='weight', node_size='size',
               edge_weight='weight', tpwgts=None, ubvec=None, options=None,
               recursive=False):
@@ -208,7 +209,7 @@ def partition(G, nparts, node_weight='weight', node_size='size',
         tpwgts = list(chain.from_iterable(tpwgts))
 
     with _zero_numbering(options):
-        objval, part = part_graph(xadj, adjncy, nparts, vwgt, vsize, adjwgt,
+        objval, part = _metis.part_graph(xadj, adjncy, nparts, vwgt, vsize, adjwgt,
                                   tpwgts, ubvec, options, recursive)
 
     parts = [[] for i in range(nparts)]
@@ -218,7 +219,7 @@ def partition(G, nparts, node_weight='weight', node_size='size',
     return objval, parts
 
 
-@convert_exceptions(nx.NetworkXError, (ValueError, TypeError, MetisError))
+@decorators.convert_exceptions(nx.NetworkXError, (ValueError, TypeError, _types.MetisError))
 def vertex_separator(G, weight='weight', options=None):
     """Compute a vertex separator that bisects a graph. The algorithm aims to
     minimize the sum of weights of vertices in the separator.
@@ -259,7 +260,7 @@ def vertex_separator(G, weight='weight', options=None):
     xadj, adjncy = _convert_graph(G)
 
     with _zero_numbering(options):
-        part = compute_vertex_separator(xadj, adjncy, vwgt, options)[1]
+        part = _metis.compute_vertex_separator(xadj, adjncy, vwgt, options)[1]
 
     groups = [[], [], []]
     for u, i in zip(G, part):
```

File: metis.py [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+from sys import exc_info
+
+import networkx as nx
+from networkx.utils import not_implemented_for
+
+import decorators
+import _types
+import _metis
+__all__ = ['node_nested_dissection', 'partition', 'vertex_separator',
+           '_types.MetisOptions']
```
Comment (ysitu) [inline metis.py]: You can bring `MetisOptions` into scope by putting

```python
MetisOption = _types.MetisOption
```

in the global scope. This is a frequently used technique.

File: metis.py [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+    finally:
+        if options:
+            options.numbering = numbering
+
+
+def _convert_graph(G):
+    """Convert a graph to the numbered adjacency list structure expected by
+    METIS.
+    """
+    index = dict(list(zip(G, list(range(len(G))))))
```
Comment (ysitu) [inline metis.py]: Not sure why I did this, but

```python
index = dict(zip(G, range(len(G))))
```

makes better sense.

File: metis.py [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+
+import decorators
+import _types
+import _metis
+__all__ = ['node_nested_dissection', 'partition', 'vertex_separator',
+           '_types.MetisOptions']
+
+
+@contextmanager
+def _zero_numbering(options):
```
Comment (ysitu) [inline metis.py]: Add a docstring

```python
"""Temporarily force zero-based numbering."""
```

File: metis.py [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+        METIS returns an error status.
+    """
+    if len(G) == 0:
+        return []
+
+    vwgt = [G.node[u].get(weight, 1) for u in G]
+    if all(w == 1 for w in vwgt):
+        vwgt = None
+
+    if G.is_directed() or G.is_multigraph():
```
Comment (ysitu) [inline metis.py]: Can we ignore `G.is_multigraph()`?

File: metis.py (hunk 12 above)
Comment (OrkoHunter) [inline metis.py]: Actually this was modified by `2to3`, you already had `index = dict(zip(G, range(len(G))))`.
`2to3` seems to make this change but it's not making sense. I'll change it.

File: metis.py [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+    index = dict(list(zip(G, list(range(len(G))))))
+    xadj = [0]
+    adjncy = []
+    for u in G:
+        adjncy.extend(index[v] for v in G[u])
+        xadj.append(len(adjncy))
+    return xadj, adjncy
+
+
+@decorators.convert_exceptions(nx.NetworkXError, (ValueError, TypeError, _types.MetisError))
```
Comment (ysitu) [inline metis.py]: Move `convert_exceptions` into this module and rename it `_convert_exceptions`.

File: metis.py (hunk 14 above)
Comment (OrkoHunter) [inline metis.py]: I don't think `partition()` is implemented for either directed graph or multi graph so, maybe we can ignore it here.

File: metis.py (hunk 15 above)
Comment (OrkoHunter) [inline metis.py]: Done.

Regarding the separate `_exec` function.
```py
def _exec(code, gvars, lvars):
    exec(code, gvars, lvars)
```
It seems that `exec` function can not be used in a function which has nested function in it with local variables. So, there's a way to use `exec` as 
```py
exec() in globals(), locals()
```
Or making a dummy function `_exec` and do the same job.
Should we go with `_exec` or use it as `exec() in globals(), locals()`?

File: metis.py [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+    adjncy = []
+    for u in G:
+        adjncy.extend(index[v] for v in G[u])
+        xadj.append(len(adjncy))
+    return xadj, adjncy
+
+def _exec(code, gvars, lvars):
+    exec(code, gvars, lvars)
+
+def _convert_exceptions(convert_type, catch_types=None):
```
Comment (ysitu) [inline metis.py]: Use two blank lines between top-level function definitions.

Comment (OrkoHunter) [inline metis.py]: Added a line here and three other places.

Himanshu Mishra

File: metis.py [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+    xadj = [0]
+    adjncy = []
+    for u in G:
+        adjncy.extend(index[v] for v in G[u])
+        xadj.append(len(adjncy))
+    return xadj, adjncy
+
+
+def _exec(code, gvars, lvars):
+    exec(code, gvars, lvars)
```
Comment (ysitu) [inline metis.py]: Document why we need this function as requested by @chebee7i.

Comment (OrkoHunter) [inline metis.py]: Yeah sure, but first please have a look at [this comment](https://github.com/networkx/networkx-metis/pull/3#discussion_r32720638). We don't actually need this function.

**Comment by chebee7i** [comment]
Are you guys depending on `six`. It looks like it provides `reraise` which let's you do this without having to code up `_exec`.

http://stackoverflow.com/a/14505093

File: metis.py (hunk 15 above)
Comment (ysitu) [inline metis.py]: Is this compatible with Python 3?

**Comment by ysitu** [comment]
@chebee7i That is actually a good suggestion. Using `six` in this case is certainly a big plus in code readability.

File: metis.py (hunk 14 above)
Comment (ysitu) [inline metis.py]: Actually I meant outright banning digraphs and multigraphs with `@not_implemented_for`.

Comment (OrkoHunter) [inline metis.py]: Okay cool. I put `@not_implemented_for` for both graphs.

**Comment by OrkoHunter** [comment]
@chebee7i Thanks for the suggestion. `six` seems to solve our problem in one another way.
The `exec` statement was replaced with an `exec(...)` function in Python 3. Replace:
```
exec code in globals, locals
```
With:
```
import six
six.exec_(code, globals, locals)
```
Source : https://github.com/ipython/ipython/wiki/IPEP-4:-Python-3-Compatibility#syntax-changes

Should I remove `_exec` and use it?

File: metis.py [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+        perm = _metis.node_nd(xadj, adjncy, vwgt, options)[0]
+
+    nodes = list(G)
+    perm = [nodes[i] for i in perm]
+
+    return perm
+
+
+@nx.utils.decorators.not_implemented_for('directed')
+@nx.utils.decorators.not_implemented_for('multigraph')
```
Comment (SanketDG) [inline metis.py]: Could be just:
```python
@nx.utils.not_implemented_for('directed')
@nx.utils.not_implemented_for('multigraph')
```
Or better, import once and use many times:
```python
from networkx.utils import not_implemented_for

@not_implemented_for('directed')
.....
```

Comment (OrkoHunter) [inline metis.py]: I guess, for
```py
@nx.utils.not_implemented_for('directed')
```
this will be needed
```py
from networkx.utils import not_implemented_for
```
The later implementation was there earlier which seem to be very tempting too but for that we'd be going against the standard import style which suggests you do not import names from other modules into yours, rather import module and then use `module.name`.

However, this seems to be a better way. Declaring a local name
```
not_implemented_for = nx.utils.decorators.not_implemented_for
```
and using it at places.

Comment (SanketDG) [inline metis.py]: > I guess, for
> 
> @nx.utils.not_implemented_for('directed')
> this will be needed
> 
> from networkx.utils import not_implemented_for

I think `networkx.utils` is imported in the top-level namespace. So you don't need the import statement.

> The later implementation was there earlier which seem to be very tempting too but for that we'd be going against the standard import style which suggests you do not import names from other modules into yours, rather import module and then use module.name.
> 
> However, this seems to be a better way. Declaring a local name
> 
> not_implemented_for = nx.utils.decorators.not_implemented_for
> and using it at places.

Yes, this is okay, but just:
```python
not_implemented_for = nx.utils.not_implemented_for
```
would suffice since `not_implemented_for` is imported in `networkx.utils`

Comment (OrkoHunter) [inline metis.py]: Yes this is right. Thanks.

Comment (OrkoHunter) [inline metis.py]: Changed it to `@nx.utils.not_implemented_for`. Probably this is more readable too.

**Comment by chebee7i** [comment]
Doesn't reraise do exactly what you are intending with those lines `_convert_exceptions`? If I understand it correctly, you won't need to call `six.exec_`, `six.reraise` will handle that for you.

**Comment by OrkoHunter** [comment]
So, if we are not using `six.exec_(code, globals, locals)`, then I'm a little unclear about the syntax with `six.raise()`. What would that be?

**Comment by chebee7i** [comment]
I haven't really looked into it, but my first guess is that something like the following could work:

```python
   def _convert_exceptions(func, *args, **kwargs):
       try:
           return func(*args, **kwargs)
       except catch_types or () as e:
           exc = sys.exec_info()
       except Exception as e:
           if catch_types is not None:
               raise
           exc = sys.exec_info()
       six.reraise(convert_type, convert_type(exc[1]), exc[2])
   return _convert_exceptions
```

**Comment by OrkoHunter** [comment]
@chebee7i  That seems to work just fine. Thanks!

**Comment by chebee7i** [comment]
It imports/compiles or it passes all the tests? :)

**Comment by OrkoHunter** [comment]
@chebee7i Yes it compiles. Tests are not passing anyway.

File: metis.py [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+
+import contextlib
+import itertools
+import sys
+import decorator
+
+import networkx as nx
+
+import _types
+import _metis
```
Comment (ysitu) [inline metis.py]: Shall we move to use absolute imports now?

Comment (OrkoHunter) [inline metis.py]: I didn't get it. I though this was preferred.

Comment (ysitu) [inline metis.py]: Sorry for being vague. I meant something like `from networkx.addons.metis import _types`.

Comment (OrkoHunter) [inline metis.py]: Doing this might force a user to install `networkx-metis` into their system before using it. In the current way, I think the user is free to install as well as test locally without installing `networkx-metis` which I personally think, is a great advantage.
Do you think the current system might cause some problem ahead as we are pass with relative imports?

Comment (chebee7i) [inline metis.py]: As far as I understand, you cannot do imports like `import _metis` in Python 3. All imports are read from sys.path and are treated as absolute (e.g. the local directory is not searched unless it is explicitly added). See here: https://www.python.org/dev/peps/pep-0328. This is precisely what we were discussing in another thread.

Comment (chebee7i) [inline metis.py]: In other words....  unless you have a dot such as `from . import _metis` then it is considered an absolute import from sys.path. Since we can't have `__init__.py` files in the namespace package, we can't do relative imports anyway.

Comment (ysitu) [inline metis.py]: I don't understand the worry about the user having to install `networkx-metis`. Installation is required anyhow---you need to build the Cython extensions.

Comment (OrkoHunter) [inline metis.py]: Yeah this is right. We surely need to change them. Updating.

Comment (OrkoHunter) [inline metis.py]: Updated `_types.py` and `metis.py` with the new import style.

**Comment by OrkoHunter** [comment]
Updated with the required directory structure.
```
networkx/
     __init__.py
    addons/
        __init.py
        metis/
            metis.py
            _types.py
            enums.py
            tests/
                test_metis.py
```

**Comment by chebee7i** [comment]
Is it still necessary to have it named _types? Absolute imports the default and in 2.7 you should be using `from __future__ import absolute_import`.

**Comment by OrkoHunter** [comment]
@chebee7i No, we can now keep both the names as well. Do you prefer changing it back to `types.py`?

**Comment by chebee7i** [comment]
I don't know what the reason for having an underscore should be, now that there is no conflict. So it seems like it should go back to `types.py`.

File: networkx/addons/metis/_types.py [hunk 20]
```diff
@@ -0,0 +1,187 @@
+import numbers
+
+import enums
+from networkx.addons.metis import _metis
+
+__all__ = ['MetisOptions', 'MetisError']
+
+
+class MetisOptions(object):
```
Comment (ysitu) [inline networkx/addons/metis/_types.py]: Add docstrings to class and properties.

File: networkx/addons/metis/_types.py [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+        names = ['ptype', 'objtype', 'ctype', 'iptype', 'rtype', 'ncuts',
+                 'nseps', 'numbering', 'niter', 'seed', 'minconn', 'no2hop',
+                 'contig', 'compress', 'ccorder', 'pfactor', 'ufactor']
+        return '{0}({1})'.format(
+            self.__class__.__name__,
+            ', '.join('{0}={1}'.format(name, repr(getattr(self, name)))
+                      for name in names))
+
+
+class MetisError(Exception):
```
Comment (ysitu) [inline networkx/addons/metis/_types.py]: Add docstrings.

File: networkx/addons/metis/enums.py [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+class MetisPType(enum.IntEnum):
+    """Partitioning method.
+
+        ==== =================================
+        rb   Multilevel recursive bisectioning
+        kway Multilevel `k`-way partitioning
+        ==== =================================
+    """
+
+    default, rb, kway = list(range(-1, 2))
```
Comment (ysitu) [inline networkx/addons/metis/enums.py]: `list()` is unnecessary.

File: networkx/addons/metis/metis.py [hunk 23]
```diff
@@ -0,0 +1,309 @@
...
+# All rights reserved
+
+"""
+Wrappers of METIS graph partitioning functions.
+"""
+
+import contextlib
+import itertools
+import sys
+import decorator
```
Comment (ysitu) [inline networkx/addons/metis/metis.py]: Order these alphabetically.

File: networkx/addons/metis/metis.py [hunk 24]
```diff
@@ -0,0 +1,309 @@
...
+
+import networkx as nx
+
+from networkx.addons.metis import _types
+from networkx.addons.metis import _metis
```
Comment (ysitu) [inline networkx/addons/metis/metis.py]: Order these too.

**Comment by ysitu** [comment]
Please address the comments about docstrings so that we can merge this PR and unblock some others. For the most part, this should involve only duplicating some text from `enums.py` into `types.py`.

**Comment by OrkoHunter** [comment]
@ysitu On it.

**Comment by OrkoHunter** [comment]
I should rename `_types` back to `types` right?

**Comment by OrkoHunter** [comment]
Docstrings added.

File: networkx/addons/metis/_types.py [hunk 25]
```diff
@@ -0,0 +1,206 @@
...
+
+
+class MetisOptions(object):
+    """Assign default values to a MetisOptions object."""
+    def __init__(self):
+        _metis.set_default_options(self)
+
+    @property
+    def ptype(self):
+        """Types of Partitioning method"""
```
Comment (ysitu) [inline networkx/addons/metis/_types.py]: I think that it is more useful to the user when they say `help(MetisOptions.ptype)` they get a list of valid enums instead of just two words. You should copy the entire docstring block here.

Comment (OrkoHunter) [inline networkx/addons/metis/_types.py]: That's nice.

File: networkx/addons/metis/_types.py [hunk 26]
```diff
@@ -0,0 +1,206 @@
+import numbers
+
+import enums
+from networkx.addons.metis import _metis
+
+__all__ = ['MetisOptions', 'MetisError']
+
+
+class MetisOptions(object):
+    """Assign default values to a MetisOptions object."""
```
Comment (ysitu) [inline networkx/addons/metis/_types.py]: Should be `Options controlling behaviors of METIS algorithms.`. Also add a blank line before `def __init__()`.

File: networkx/addons/metis/_types.py [hunk 27]
```diff
@@ -0,0 +1,206 @@
...
+    def __init__(self):
+        _metis.set_default_options(self)
```
Comment (ysitu) [inline networkx/addons/metis/_types.py]: This also needs a docstring. `Initializes a MetisOptions object.` should be good enough.

BTW, it is probably useful to enable this syntax:

```python
options = MetisOptions(niter=100)
```

so that you can avoid the verbosity in

```python
options = MetisOptions()
options.niter = 100
```

An extra `**kwargs` parameter should do the trick. You need to extend the docstring for that, though.

File: networkx/addons/metis/_types.py [hunk 28]
```diff
@@ -0,0 +1,272 @@
...
+import enums
+from networkx.addons.metis import _metis
+
+__all__ = ['MetisOptions', 'MetisError']
+
+
+class MetisOptions(object):
+    """Options controlling behaviors of METIS algorithms"""
+
+    def __init__(self, **kwargs):
```
Comment (OrkoHunter) [inline networkx/addons/metis/_types.py]: Updated with `**kwargs` option and docstring with one example.

File: networkx/addons/metis/_types.py [hunk 29]
```diff
@@ -0,0 +1,272 @@
...
+        """Initializes a MetisOptions object. Values can be provided
+        for some parameters as arguments.
+
+        Example
+        -------
+        >>> options = MetisOptions(ncuts=2, niter=100)
+
+        """
+        _metis.set_default_options(self)
```
Comment (ysitu) [inline networkx/addons/metis/_types.py]: You need to do something with `kwargs`, right?

File: networkx/addons/metis/_types.py [hunk 30]
```diff
@@ -0,0 +1,272 @@
...
+    def __init__(self, **kwargs):
+        """Initializes a MetisOptions object. Values can be provided
+        for some parameters as arguments.
+
+        Example
+        -------
+        >>> options = MetisOptions(ncuts=2, niter=100)
+
+        """
+        _metis.set_default_options(self, **kwargs)
```
Comment (OrkoHunter) [inline networkx/addons/metis/_types.py]: Will this do?

Comment (ysitu) [inline networkx/addons/metis/_types.py]: No. You cannot just try things randomly. You need to think about what things would work and why. In this case, when the user says `MetisOptions(niter=100)`, you want to achieve the effect of `options.niter = 100`. You need to find out where to extract `niter` and `100` and how to set them on `options`.

File: networkx/addons/metis/_types.py [hunk 31]
```diff
@@ -0,0 +1,275 @@
...
+        Example
+        -------
+        >>> options = MetisOptions(ncuts=2, niter=100)
+
+        """
+        _metis.set_default_options(self)
+        for key, value in kwargs.items():
+            if (value != None):
+                setattr(self, '_' + key, value)
+
```
Comment (OrkoHunter) [inline networkx/addons/metis/_types.py]: This one should go. It works.

Comment (ysitu) [inline networkx/addons/metis/_types.py]: You should not add an underscore, otherwise the validation checks will be bypassed.

File: networkx/addons/metis/_types.py [hunk 32]
```diff
@@ -0,0 +1,275 @@
...
+        for some parameters as arguments.
+
+        Example
+        -------
+        >>> options = MetisOptions(ncuts=2, niter=100)
+
+        """
+        _metis.set_default_options(self)
+        for key, value in kwargs.items():
+            if (value != None):
```
Comment (ysitu) [inline networkx/addons/metis/_types.py]: Remove parentheses. Always use `is not None` instead of `!= None`.

File: networkx/addons/metis/tests/test_metis.py [hunk 33]
```diff
@@ -0,0 +1,70 @@
...
+        adjncy = list(
+            itertools.chain.from_iterable(
+                zip(itertools.chain([n - 1], range(n - 1)),
+                    range(n),
+                    itertools.chain(range(1, n), [0]))))
+        perm, iperm = _metis.node_nd(xadj, adjncy)
+        nose.tools.assert_equal(set(perm), set(range(n)))
+        nose.tools.assert_equal(abs(perm[-1] - perm[-2]), n // 2)
+        nose.tools.ok_(set(range(min(perm[-2:]) + 1, max(perm[-2:]))) in
+                      (set(perm[0:n // 2 - 1]), set(perm[n // 2 - 1:-2])))
```
Comment (ysitu) [inline networkx/addons/metis/tests/test_metis.py]: One extra space of indentation.

File: networkx/addons/metis/tests/test_metis.py [hunk 34]
```diff
@@ -0,0 +1,70 @@
...
+class TestMetis(object):
+
+    def test_node_nd(self):
+        n = 16
+        xadj, adjncy = make_cycle(n)
+        perm, iperm = _metis.node_nd(xadj, adjncy)
+        nose.tools.assert_equal(set(perm), set(range(n)))
+        nose.tools.assert_equal(abs(perm[-1] - perm[-2]), n // 2)
+        nose.tools.ok_(set(range(min(perm[-2:]) + 1, max(perm[-2:]))) in
+                      (set(perm[0:n // 2 - 1]), set(perm[n // 2 - 1:-2])))
```
Comment (ysitu) [inline networkx/addons/metis/tests/test_metis.py]: One extra space of indentation.

**Comment by ysitu** [comment]
> I should rename _types back to types right?

With absolute imports now it makes sense to rename.

**Comment by ysitu** [comment]
I will merge this after the last three comments are addressed.

**Comment by OrkoHunter** [comment]
> I will merge this after the last three comments are addressed.

Shouldn't I add the https://github.com/networkx/networkx-metis/pull/14#issuecomment-115105652 about error raised in setting the value of `niter` in this PR?

**Comment by ysitu** [comment]
You can keep the new test in #14.

**Comment by OrkoHunter** [comment]
Indentations corrected. `_types` renamed back to `types` and edits made in `metis.py`

**Comment by OrkoHunter** [comment]
There's a [line](https://github.com/networkx/networkx-metis/blob/master/_metis.pyx#L7) in `_metis.pyx` where `_types` is being imported. How do you think should I make that change? With a commit in #14 or a new PR?

**Comment by ysitu** [comment]
It is okay to piggyback that on #14.

**Comment by OrkoHunter** [comment]
Okay. Thanks for merging!
